File: desktop/native/SurveyController.App/Views/StrategyEditor.Tree.cpp
#include "pch.h"
#include "StrategyEditor.xaml.h"
#include "Services/JsonHelpers.h"

#include <algorithm>
#include <map>
#include <cwctype>
#include <tuple>

namespace winrt::SurveyController::App::implementation
{
    namespace
    {
        using namespace Microsoft::UI::Xaml;
        using namespace Microsoft::UI::Xaml::Controls;
        using namespace Windows::Data::Json;

        hstring ShortTitle(hstring const& value)
        {
            std::wstring title{ value.empty() ? L"未命名题目" : value.c_str() };
            constexpr size_t limit = 24;
            if (title.size() > limit) title = title.substr(0, limit - 1) + L"…";
            return hstring{ title };
        }

        int32_t PageNumber(JsonObject const& question)
        {
            return (std::max)(1, static_cast<int32_t>(question.GetNamedNumber(L"page", 1)));
        }

        struct SearchMatch
        {
            int32_t count{};
            hstring firstLabel;
        };

        SearchMatch MatchSearchSegments(JsonObject const& question, hstring const& query)
        {
            std::wstring needle{ query };
            std::transform(needle.begin(), needle.end(), needle.begin(), ::towlower);
            if (needle.empty()) return {};

            SearchMatch match;
            for (auto const& value : Services::GetJsonArray(question, L"search_segments"))
            {
                if (value.ValueType() != JsonValueType::Object) continue;
                auto segment = value.GetObject();
                auto label = segment.GetNamedString(L"label", L"内容");
                std::wstring haystack{ label };
                haystack += L" " + std::wstring{ segment.GetNamedString(L"text", L"") };
                std::transform(haystack.begin(), haystack.end(), haystack.begin(), ::towlower);
                if (haystack.find(needle) == std::wstring::npos) continue;
                if (match.firstLabel.empty()) match.firstLabel = label;
                ++match.count;
            }
            return match;
        }

        hstring SearchPositionText(std::vector<int32_t> const& matches, int32_t currentIndex, int32_t occurrences)
        {
            if (matches.empty()) return L"0 题匹配";
            auto current = std::find(matches.begin(), matches.end(), currentIndex);
            auto position = current == matches.end() ? 0 : static_cast<int32_t>(std::distance(matches.begin(), current) + 1);
            return hstring{ std::to_wstring(position) + L" / " + std::to_wstring(matches.size()) + L" 题 · "
                + std::to_wstring(occurrences) + L" 处匹配" };
        }

        UIElement QuestionNodeContent(Services::WizardQuestion const& question)
        {
            auto title = TextBlock{};
            title.Text(hstring{ std::to_wstring(question.number) + L". " + std::wstring{ ShortTitle(question.title) } });
            title.TextTrimming(TextTrimming::CharacterEllipsis);
            Automation::AutomationProperties::SetName(title, hstring{ L"第 " + std::to_wstring(question.number) + L" 题，" +
                std::wstring{ question.type } + L"，" + std::wstring{ question.title } });
            Automation::AutomationProperties::SetAutomationId(title, hstring{ L"AnswerEditor.Question." + std::to_wstring(question.number) });
            return title;
        }

        UIElement PageNodeContent(int32_t page, int32_t questionCount)
        {
            auto text = TextBlock{};
            text.Text(hstring{ L"第 " + std::to_wstring(page) + L" 页 · " + std::to_wstring(questionCount) + L" 题" });
            Automation::AutomationProperties::SetName(text, text.Text());
            Automation::AutomationProperties::SetAutomationId(text, hstring{ L"AnswerEditor.Page." + std::to_wstring(page) });
            return text;
        }

        UIElement RelationNodeContent(JsonObject const& relation)
        {
            auto label = TextBlock{};
            label.Text(relation.GetNamedString(L"label", L"逻辑关系"));
            label.FontWeight(Windows::UI::Text::FontWeights::SemiBold());
            label.TextTrimming(TextTrimming::CharacterEllipsis);
            auto summary = relation.GetNamedString(L"summary", L"");
            Automation::AutomationProperties::SetName(label, summary);
            Automation::AutomationProperties::SetAutomationId(label, relation.GetNamedString(L"id", L"AnswerEditor.Relation"));
            return label;
        }

        StackPanel SearchSuggestion(Services::WizardQuestion const& question, SearchMatch const& match)
        {
            auto panel = StackPanel{};
            panel.Spacing(1);
            panel.Tag(box_value(question.number));
            auto primary = TextBlock{};
            primary.Text(hstring{ L"第 " + std::to_wstring(question.number) + L" 题 · " + std::wstring{ ShortTitle(question.title) } });
            auto secondary = TextBlock{};
            secondary.Text(hstring{ std::wstring{ match.firstLabel } + L" · " + std::to_wstring(match.count) + L" 处匹配" });
            secondary.FontSize(12);
            secondary.Foreground(Application::Current().Resources().Lookup(box_value(L"TextFillColorSecondaryBrush")).as<Microsoft::UI::Xaml::Media::Brush>());
            panel.Children().Append(primary);
            panel.Children().Append(secondary);
            Automation::AutomationProperties::SetName(panel, hstring{ std::wstring{ primary.Text() } + L"，" + std::wstring{ secondary.Text() } });
            Automation::AutomationProperties::SetAutomationId(panel, hstring{ L"AnswerEditor.SearchResult." + std::to_wstring(question.number) });
            return panel;
        }

        int32_t SuggestionQuestionNumber(IInspectable const& value)
        {
            auto element = value.try_as<FrameworkElement>();
            return element ? unbox_value_or<int32_t>(element.Tag(), 0) : 0;
        }
    }

    void StrategyEditor::RebuildQuestionTree(int32_t selectedIndex)
    {
        auto tree = QuestionTree();
        tree.RootNodes().Clear();
        m_questionNodes.clear();
        m_treeTargets.clear();

        auto questions = m_document.Questions();
        m_questionNodes.resize(questions.size());
        if (questions.empty())
        {
            m_questionIndex = -1;
            return;
        }

        std::map<int32_t, TreeViewNode> pages;
        auto pageNode = [&](int32_t page, int32_t questionCount)
        {
            auto found = pages.find(page);
            if (found != pages.end()) return found->second;
            TreeViewNode node;
            node.Content(PageNodeContent(page, questionCount));
            node.IsExpanded(true);
            tree.RootNodes().Append(node);
            pages.emplace(page, node);
            return node;
        };

        for (uint32_t index = 0; index < questions.size(); ++index)
        {
            auto const& summary = questions[index];
            auto question = m_document.QuestionAt(index);
            auto parent = pageNode(PageNumber(question), summary.pageQuestionCount);

            TreeViewNode node;
            node.Content(QuestionNodeContent(summary));
            parent.Children().Append(node);
            m_questionNodes[index] = node;
            m_treeTargets.emplace_back(node, static_cast<int32_t>(index));

            auto appendRelations = [&](JsonArray const& relations, bool inbound)
            {
                for (auto const& value : relations)
                {
                    if (value.ValueType() != JsonValueType::Object) continue;
                    auto relation = value.GetObject();
                    auto targetNumber = static_cast<int32_t>(relation.GetNamedNumber(
                        inbound ? L"sourceQuestionNum" : L"targetQuestionNum", 0));
                    if (targetNumber <= 0) continue;
                    auto target = -1;
                    for (uint32_t candidate = 0; candidate < questions.size(); ++candidate)
                    {
                        if (questions[candidate].number == targetNumber)
                        {
                            target = static_cast<int32_t>(candidate);
                            break;
                        }
                    }
                    if (target < 0) continue;
                    TreeViewNode relationNode;
                    relationNode.Content(RelationNodeContent(relation));
                    // Keep relations as first-class page entries. Nesting them
                    // under every question makes the tree tall and hides the
                    // actual navigation targets behind expansion state.
                    parent.Children().Append(relationNode);
                    m_treeTargets.emplace_back(relationNode, target);
                }
            };
            appendRelations(Services::GetJsonArray(question, L"inbound_relations"), true);
            appendRelations(Services::GetJsonArray(question, L"outbound_relations"), false);
        }

        if (tree.RootNodes().Size() == 0)
        {
            m_questionIndex = -1;
            QuestionTitle().Text(L"没有匹配的题目");
            QuestionMeta().Text(L"清空搜索框后显示全部题目");
            return;
        }
        auto validIndex = selectedIndex >= 0 && selectedIndex < static_cast<int32_t>(questions.size())
            && m_questionNodes[static_cast<size_t>(selectedIndex)]
            ? selectedIndex : -1;
        if (validIndex < 0)
        {
            for (size_t index = 0; index < m_questionNodes.size(); ++index)
            {
                if (m_questionNodes[index]) { validIndex = static_cast<int32_t>(index); break; }
            }
        }
        SelectQuestion(validIndex);
    }

    void StrategyEditor::SelectQuestion(int32_t index)
    {
        if (index < 0 || index >= static_cast<int32_t>(m_questionNodes.size())) return;
        if (!m_questionNodes[static_cast<size_t>(index)]) return;
        if (m_questionIndex >= 0 && m_questionIndex != index && !SaveCurrentQuestion())
        {
            auto previous = m_questionNodes[static_cast<size_t>(m_questionIndex)];
            m_syncingTreeSelection = true;
            QuestionTree().SelectedNode(previous);
            m_syncingTreeSelection = false;
            return;
        }
        m_questionIndex = index;
        auto node = m_questionNodes[static_cast<size_t>(index)];
        if (QuestionTree().SelectedNode() != node)
        {
            m_syncingTreeSelection = true;
            QuestionTree().SelectedNode(node);
            m_syncingTreeSelection = false;
        }
        LoadQuestion();
        if (ActualWidth() < 800)
        {
            QuestionListPane().Visibility(Visibility::Collapsed);
            QuestionDetailScroll().Visibility(Visibility::Visible);
            BackToListButton().Visibility(Visibility::Visible);
        }
    }

    void StrategyEditor::OnQuestionSearchChanged(IInspectable const&, AutoSuggestBoxTextChangedEventArgs const& args)
    {
        if (args.Reason() == AutoSuggestionBoxTextChangeReason::UserInput)
        {
            auto suggestions = winrt::single_threaded_observable_vector<IInspectable>();
            auto questions = m_document.Questions();
            std::vector<int32_t> matches;
            int32_t occurrences = 0;
            for (uint32_t index = 0; index < questions.size(); ++index)
            {
                auto match = MatchSearchSegments(m_document.QuestionAt(index), QuestionSearch().Text());
                if (match.count == 0) continue;
                matches.push_back(static_cast<int32_t>(index));
                occurrences += match.count;
                if (suggestions.Size() < 12) suggestions.Append(SearchSuggestion(questions[index], match));
            }
            QuestionSearch().ItemsSource(suggestions);
            SearchPosition().Text(QuestionSearch().Text().empty() ? L"" : SearchPositionText(matches, m_questionIndex, occurrences));
        }
    }

    void StrategyEditor::OnQuestionSuggestionChosen(IInspectable const&, AutoSuggestBoxSuggestionChosenEventArgs const& args)
    {
        auto number = SuggestionQuestionNumber(args.SelectedItem());
        if (number <= 0) return;
        auto questions = m_document.Questions();
        for (uint32_t index = 0; index < questions.size(); ++index)
        {
            if (questions[index].number == number) { SelectQuestion(static_cast<int32_t>(index)); return; }
        }
    }

    void StrategyEditor::OnQuestionQuerySubmitted(IInspectable const&, AutoSuggestBoxQuerySubmittedEventArgs const& args)
    {
        auto chosenNumber = args.ChosenSuggestion() ? SuggestionQuestionNumber(args.ChosenSuggestion()) : 0;
        auto questions = m_document.Questions();
        std::vector<int32_t> matches;
        int32_t occurrences = 0;
        for (uint32_t index = 0; index < questions.size(); ++index)
        {
            auto match = MatchSearchSegments(m_document.QuestionAt(index), args.QueryText());
            if (match.count == 0) continue;
            matches.push_back(static_cast<int32_t>(index));
            occurrences += match.count;
        }
        if (chosenNumber > 0)
        {
            for (uint32_t index = 0; index < questions.size(); ++index)
            {
                if (questions[index].number == chosenNumber)
                {
                    SelectQuestion(static_cast<int32_t>(index));
                    SearchPosition().Text(SearchPositionText(matches, m_questionIndex, occurrences));
                    return;
                }
            }
            return;
        }
        if (!matches.empty())
        {
            SelectQuestion(matches.front());
            SearchPosition().Text(SearchPositionText(matches, m_questionIndex, occurrences));
        }
    }

    void StrategyEditor::OnQuestionSelected(IInspectable const&, TreeViewSelectionChangedEventArgs const&)
    {
        if (m_syncingTreeSelection) return;
        auto node = QuestionTree().SelectedNode();
        if (!node) return;
        for (auto const& [candidate, targetIndex] : m_treeTargets)
        {
            if (candidate == node)
            {
                auto lifetime = get_strong();
                DispatcherQueue().TryEnqueue([lifetime, targetIndex]()
                {
                    if (targetIndex >= 0 && targetIndex < static_cast<int32_t>(lifetime->m_questionNodes.size()))
                    {
                        try { lifetime->SelectQuestion(targetIndex); }
                        catch (hresult_error const& error)
                        {
                            lifetime->QuestionStatus().Severity(InfoBarSeverity::Error);
                            lifetime->QuestionStatus().Title(L"切换题目失败");
                            lifetime->QuestionStatus().Message(error.message());
                            lifetime->QuestionStatus().IsOpen(true);
                        }
                        catch (...)
                        {
                            lifetime->QuestionStatus().Severity(InfoBarSeverity::Error);
                            lifetime->QuestionStatus().Title(L"切换题目失败");
                            lifetime->QuestionStatus().Message(L"题目状态已更新，请重新选择。");
                            lifetime->QuestionStatus().IsOpen(true);
                        }
                    }
                });
                return;
            }
        }
    }

    void StrategyEditor::OnQuestionInvoked(IInspectable const&, TreeViewItemInvokedEventArgs const& args)
    {
        auto node = args.InvokedItem().try_as<TreeViewNode>();
        if (!node) return;
        for (auto const& [candidate, targetIndex] : m_treeTargets)
        {
            if (candidate == node)
            {
                auto lifetime = get_strong();
                DispatcherQueue().TryEnqueue([lifetime, targetIndex]()
                {
                    if (targetIndex < 0 || targetIndex >= static_cast<int32_t>(lifetime->m_questionNodes.size())) return;
                    try { lifetime->SelectQuestion(targetIndex); }
                    catch (...)
                    {
                        lifetime->QuestionStatus().Severity(InfoBarSeverity::Error);
                        lifetime->QuestionStatus().Title(L"切换题目失败");
                        lifetime->QuestionStatus().Message(L"题目状态已更新，请重新选择。");
                        lifetime->QuestionStatus().IsOpen(true);
                    }
                });
                return;
            }
        }
    }

    void StrategyEditor::SelectNextMatch()
    {
        auto questions = m_document.Questions();
        std::vector<int32_t> matches;
        int32_t occurrences = 0;
        for (uint32_t index = 0; index < questions.size(); ++index)
        {
            auto match = MatchSearchSegments(m_document.QuestionAt(index), QuestionSearch().Text());
            if (match.count == 0) continue;
            matches.push_back(static_cast<int32_t>(index));
            occurrences += match.count;
        }
        if (matches.empty()) return;
        auto position = std::find(matches.begin(), matches.end(), m_questionIndex);
        auto next = position == matches.end() || ++position == matches.end() ? matches.begin() : position;
        SelectQuestion(*next);
        SearchPosition().Text(SearchPositionText(matches, m_questionIndex, occurrences));
    }

    void StrategyEditor::SelectPreviousMatch()
    {
        auto questions = m_document.Questions();
        std::vector<int32_t> matches;
        int32_t occurrences = 0;
        for (uint32_t index = 0; index < questions.size(); ++index)
        {
            auto match = MatchSearchSegments(m_document.QuestionAt(index), QuestionSearch().Text());
            if (match.count == 0) continue;
            matches.push_back(static_cast<int32_t>(index));
            occurrences += match.count;
        }
        if (matches.empty()) return;
        auto position = std::find(matches.begin(), matches.end(), m_questionIndex);
        auto previous = position == matches.end() || position == matches.begin() ? std::prev(matches.end()) : std::prev(position);
        SelectQuestion(*previous);
        SearchPosition().Text(SearchPositionText(matches, m_questionIndex, occurrences));
    }

}

File: desktop/native-cs/src/SurveyController.App/Views/StrategyEditor.Tree.cs
using System.Globalization;
using System.Text.Json.Nodes;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Text;
using Microsoft.Windows.BadgeNotifications;

namespace SurveyController.App.Views;

/// <summary>题目树构建、搜索与选择，对照 C++ StrategyEditor.Tree.cpp。</summary>
public sealed partial class StrategyEditor
{
    private void RebuildQuestionTree(int selectedIndex)
    {
        QuestionTree.RootNodes.Clear();
        _questionNodes.Clear();
        _treeTargets.Clear();

        var questions = Document.Questions();
        if (questions.Count == 0)
        {
            _questionIndex = -1;
            UpdateTaskbarBadge(0);
            return;
        }

        var hasUnknownLogic = false;
        var unsupportedCount = 0;
        var maxQuestionNumber = 0;
        var questionIndices = new Dictionary<int, int>();
        for (var index = 0; index < questions.Count; index++)
        {
            var question = Document.QuestionAt(index) ?? new JsonObject();
            var status = JsonStr(question, "logic_parse_status", string.Empty);
            hasUnknownLogic |= status.ToLowerInvariant() == "unknown";
            unsupportedCount += questions[index].Unsupported ? 1 : 0;
            maxQuestionNumber = Math.Max(maxQuestionNumber, questions[index].Number);
            questionIndices.TryAdd(questions[index].Number, index);
        }

        var pages = new Dictionary<int, TreeViewNode>();
        TreeViewNode PageNode(int page)
        {
            if (pages.TryGetValue(page, out var existing))
            {
                return existing;
            }
            var node = new TreeViewNode
            {
                Content = $"第 {page} 页",
                IsExpanded = true,
            };
            QuestionTree.RootNodes.Add(node);
            pages[page] = node;
            return node;
        }

        void AppendRelation(TreeViewNode parent, string label, int targetIndex)
        {
            var node = new TreeViewNode { Content = label };
            parent.Children.Add(node);
            _treeTargets.Add((node, targetIndex));
        }

        for (var index = 0; index < questions.Count; index++)
        {
            var summary = questions[index];
            var question = Document.QuestionAt(index) ?? new JsonObject();
            var parent = PageNode(Math.Max(1, (int)JsonNumber(question, "page", 1)));

            var node = new TreeViewNode { Content = BuildQuestionNodeContent(summary) };
            parent.Children.Add(node);
            _questionNodes.Add(node);
            _treeTargets.Add((node, index));

            if (!hasUnknownLogic)
            {
                foreach (var value in JsonArrayField(question, "controls_display_targets"))
                {
                    if (value is not JsonObject target)
                    {
                        continue;
                    }
                    var targetNumber = (int)JsonNumber(target, "target_question_num", 0);
                    if (targetNumber <= 0)
                    {
                        continue;
                    }
                    var targetIndex = questionIndices.GetValueOrDefault(targetNumber, index);
                    var optionIndices = target["condition_option_indices"] as JsonArray ?? [];
                    AppendRelation(node,
                        $"条件 · 选中{OptionLabel(question, optionIndices)} → 显示第 {targetNumber} 题",
                        targetIndex);
                }

                foreach (var value in JsonArrayField(question, "jump_rules"))
                {
                    if (value is not JsonObject rule)
                    {
                        continue;
                    }
                    var targetNumber = (int)JsonNumber(rule, "jumpto", 0);
                    if (targetNumber <= 0)
                    {
                        continue;
                    }
                    var optionIndex = new JsonArray { JsonNumber(rule, "option_index", -1) };
                    var endsSurvey = JsonBool(rule, "terminates_survey") || targetNumber > maxQuestionNumber;
                    var targetIndex = !endsSurvey && questionIndices.TryGetValue(targetNumber, out var found)
                        ? found
                        : index;
                    var targetLabel = endsSurvey ? "结束" : $"第 {targetNumber} 题";
                    AppendRelation(node,
                        $"跳题 · 选中{OptionLabel(question, optionIndex)} → {targetLabel}",
                        targetIndex);
                }
            }
            node.IsExpanded = node.Children.Count > 0;
        }

        if (QuestionTree.RootNodes.Count == 0)
        {
            _questionIndex = -1;
            UpdateTaskbarBadge((uint)unsupportedCount);
            QuestionTitle.Text = "没有匹配的题目";
            QuestionMeta.Text = "清空搜索框后显示全部题目";
            QuestionCountSummary.Text = $"0 / {questions.Count} 题";
            return;
        }
        UpdateTaskbarBadge((uint)unsupportedCount);
        var validIndex = selectedIndex >= 0 && selectedIndex < questions.Count ? selectedIndex : -1;
        if (validIndex < 0)
        {
            validIndex = 0;
        }
        SelectQuestion(validIndex);
    }

    private void SelectQuestion(int index)
    {
        if (index < 0 || index >= _questionNodes.Count)
        {
            return;
        }
        if (_questionIndex >= 0 && _questionIndex != index && !SaveCurrentQuestion())
        {
            SetTreeSelection(_questionIndex);
            return;
        }
        _questionIndex = index;
        QuestionCountSummary.Text = $"第 {index + 1} / {_questionNodes.Count} 题";
        PreviousQuestionButton.IsEnabled = index > 0;
        NextQuestionButton.IsEnabled = index + 1 < _questionNodes.Count;
        SetTreeSelection(index);
        LoadQuestion();
    }

    private void SetTreeSelection(int index)
    {
        if (index < 0 || index >= _questionNodes.Count)
        {
            return;
        }
        var node = _questionNodes[index];
        if (!ReferenceEquals(QuestionTree.SelectedNode, node))
        {
            _syncingTreeSelection = true;
            QuestionTree.SelectedNode = node;
            _syncingTreeSelection = false;
        }
    }

    private void OnQuestionSearchChanged(AutoSuggestBox sender, AutoSuggestBoxTextChangedEventArgs args)
    {
        if (args.Reason != AutoSuggestionBoxTextChangeReason.UserInput)
        {
            return;
        }
        var suggestions = new List<string>();
        var questions = Document.Questions();
        for (var index = 0; index < questions.Count; index++)
        {
            var question = questions[index];
            var raw = Document.QuestionAt(index) ?? new JsonObject();
            if (!ContainsText(raw, sender.Text, question.Number))
            {
                continue;
            }
            suggestions.Add($"第 {question.Number} · {question.Type} · {ShortTitle(question.Title)}"
                + (question.LogicSummary.Length == 0 ? "" : $" · {question.LogicSummary}"));
            if (suggestions.Count >= 12)
            {
                break;
            }
        }
        sender.ItemsSource = suggestions;
    }

    private void OnQuestionSuggestionChosen(AutoSuggestBox sender, AutoSuggestBoxSuggestionChosenEventArgs args)
    {
        SelectBySuggestionText(args.SelectedItem?.ToString() ?? string.Empty);
    }

    private void OnQuestionQuerySubmitted(AutoSuggestBox sender, AutoSuggestBoxQuerySubmittedEventArgs args)
    {
        var text = args.ChosenSuggestion?.ToString() ?? args.QueryText;
        if (text.Length == 0)
        {
            return;
        }
        // 候选项形如“第 N · …”；纯文本则按内容模糊匹配第一题。
        if (text.StartsWith('第') && TrySelectByNumber(text)) 
        {
            return;
        }
        var questions = Document.Questions();
        for (var index = 0; index < questions.Count; index++)
        {
            if (ContainsText(Document.QuestionAt(index) ?? new JsonObject(), text, questions[index].Number))
            {
                SelectQuestion(index);
                return;
            }
        }
    }

    private bool TrySelectByNumber(string text)
    {
        var separator = text.IndexOf(' ');
        if (separator < 2)
        {
            return false;
        }
        if (!int.TryParse(text.Substring(1, separator - 1), out var number))
        {
            return false;
        }
        var questions = Document.Questions();
        for (var index = 0; index < questions.Count; index++)
        {
            if (questions[index].Number == number)
            {
                SelectQuestion(index);
                return true;
            }
        }
        return false;
    }

    private void SelectBySuggestionText(string selectedText)
    {
        if (selectedText.Length == 0)
        {
            return;
        }
        TrySelectByNumber(selectedText);
    }

    private void OnQuestionSelected(object sender, TreeViewSelectionChangedEventArgs e)
    {
        if (_syncingTreeSelection)
        {
            return;
        }
        var node = QuestionTree.SelectedNode;
        if (node is null)
        {
            return;
        }
        foreach (var (candidate, targetIndex) in _treeTargets)
        {
            if (ReferenceEquals(candidate, node))
            {
                _ = DispatcherQueue.TryEnqueue(() =>
                {
                    if (targetIndex >= 0 && targetIndex < _questionNodes.Count)
                    {
                        try
                        {
                            SelectQuestion(targetIndex);
                        }
                        catch (Exception error)
                        {
                            ShowSwitchError(error.Message.Length > 0 ? error.Message : "题目状态已更新，请重新选择。");
                        }
                    }
                });
                return;
            }
        }
    }

    private void OnQuestionInvoked(TreeView sender, TreeViewItemInvokedEventArgs args)
    {
        if (args.InvokedItem is not TreeViewNode node)
        {
            return;
        }
        foreach (var (candidate, targetIndex) in _treeTargets)
        {
            if (ReferenceEquals(candidate, node))
            {
                _ = DispatcherQueue.TryEnqueue(() =>
                {
                    if (targetIndex >= 0 && targetIndex < _questionNodes.Count)
                    {
                        try
                        {
                            SelectQuestion(targetIndex);
                        }
                        catch (Exception)
                        {
                            ShowSwitchError("题目状态已更新，请重新选择。");
                        }
                    }
                });
                return;
            }
        }
    }

    private void ShowSwitchError(string message)
    {
        QuestionStatus.Severity = InfoBarSeverity.Error;
        QuestionStatus.Title = "切换题目失败";
        QuestionStatus.Message = message;
        QuestionStatus.IsOpen = true;
    }

    private void OnPreviousQuestion(object sender, RoutedEventArgs e)
    {
        if (_questionIndex > 0)
        {
            SelectQuestion(_questionIndex - 1);
        }
    }

    private void OnNextQuestion(object sender, RoutedEventArgs e)
    {
        if (_questionIndex >= 0 && _questionIndex + 1 < _questionNodes.Count)
        {
            SelectQuestion(_questionIndex + 1);
        }
    }

    // —— 节点视觉 ——

    internal static string ShortTitle(string value)
    {
        var title = value.Length == 0 ? "未命名题目" : value;
        const int limit = 24;
        return title.Length > limit ? title[..(limit - 1)] + "…" : title;
    }

    private static string OptionLabel(JsonObject question, JsonArray indices)
    {
        var optionTexts = question["option_texts"] as JsonArray ?? [];
        var result = string.Empty;
        var count = 0;
        foreach (var value in indices)
        {
            if (value is not System.Text.Json.Nodes.JsonValue || !value.TryGetValue<double>(out var raw))
            {
                continue;
            }
            var index = (int)raw;
            if (index < 0)
            {
                continue;
            }
            if (count++ > 0)
            {
                result += "、";
            }
            result += index < optionTexts.Count
                ? $"“{ShortTitle(AsString(optionTexts[index]))}”"
                : $"第 {index + 1} 项";
            if (count == 3 && indices.Count > 3)
            {
                result += $"等{indices.Count}项";
                break;
            }
        }
        return result.Length == 0 ? "指定选项" : result;
    }

    private static bool ContainsText(JsonObject question, string query, int number)
    {
        var needle = query.ToLowerInvariant();
        if (needle.Length == 0)
        {
            return true;
        }
        var haystack = $"{number} {JsonStr(question, "title", string.Empty)}";
        foreach (var value in question["option_texts"] as JsonArray ?? [])
        {
            if (value is System.Text.Json.Nodes.JsonValue text)
            {
                haystack += " " + AsString(text);
            }
        }
        return haystack.ToLowerInvariant().Contains(needle, StringComparison.Ordinal);
    }

    private static (string ForegroundKey, string BadgeForegroundKey, string BackgroundKey) QuestionBrushKeys(string type) =>
        type switch
        {
            "single" or "multiple" or "dropdown" => ("QuestionChoiceBrush", "QuestionChoiceBadgeForegroundBrush", "QuestionChoiceBadgeBackgroundBrush"),
            "text" or "multi_text" or "location" => ("QuestionTextBrush", "QuestionTextBadgeForegroundBrush", "QuestionTextBadgeBackgroundBrush"),
            "scale" or "slider" => ("QuestionScaleBrush", "QuestionScaleBadgeForegroundBrush", "QuestionScaleBadgeBackgroundBrush"),
            "matrix" => ("QuestionMatrixBrush", "QuestionMatrixBadgeForegroundBrush", "QuestionMatrixBadgeBackgroundBrush"),
            "sort" => ("QuestionSortBrush", "QuestionSortBadgeForegroundBrush", "QuestionSortBadgeBackgroundBrush"),
            _ => ("QuestionUnknownBrush", "QuestionUnknownBadgeForegroundBrush", "QuestionUnknownBadgeBackgroundBrush"),
        };

    private static StackPanel BuildQuestionNodeContent(WizardQuestion question)
    {
        var (foregroundKey, badgeForegroundKey, backgroundKey) = QuestionBrushKeys(question.NormalizedType);
        var resources = Application.Current.Resources;
        var foreground = resources[foregroundKey] as Brush;
        var badgeForeground = resources[badgeForegroundKey] as Brush;
        var row = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 7 };

        row.Children.Add(new FontIcon
        {
            Glyph = question.Icon,
            FontSize = 14,
            Foreground = foreground,
        });
        row.Children.Add(new InfoBadge
        {
            Width = 8,
            Height = 8,
            VerticalAlignment = VerticalAlignment.Center,
            Background = resources[backgroundKey] as Brush,
        });
        row.Children.Add(new TextBlock
        {
            Text = question.Type,
            FontSize = 12,
            FontWeight = FontWeights.SemiBold,
            Foreground = badgeForeground,
        });
        row.Children.Add(new TextBlock
        {
            Text = $"{question.Number}. {ShortTitle(question.Title)}",
            TextTrimming = TextTrimming.CharacterEllipsis,
        });

        void AppendStatus(string label, string brushKey)
        {
            // 官方元素组合的轻量状态徽标：文本 + 主题画刷背景（替代自绘模板）。
            row.Children.Add(new TextBlock
            {
                Text = label,
                Style = Application.Current.Resources["WizardBadgeTextStyle"] as Style,
                Background = resources[brushKey] as Brush,
                CornerRadius = new CornerRadius(4),
                Padding = new Thickness(5, 1, 5, 2),
                Margin = new Thickness(2, 0, 0, 0),
            });
        }
        if (question.Required)
        {
            AppendStatus("必答", "RequiredBadgeForegroundBrush");
        }
        if (question.HasJump || question.HasDisplayLogic)
        {
            AppendStatus("逻辑", "LogicBadgeForegroundBrush");
        }
        if (question.Unsupported)
        {
            AppendStatus("不支持", "UnsupportedBadgeForegroundBrush");
        }
        AutomationProperties.SetName(row, $"第 {question.Number} 题，{question.Type}，{question.Title}");
        return row;
    }

    private static void UpdateTaskbarBadge(uint unsupportedCount)
    {
        try
        {
            var manager = BadgeNotificationManager.Current();
            if (unsupportedCount == 0)
            {
                manager.ClearBadge();
            }
            else
            {
                manager.SetBadgeAsCount(unsupportedCount);
            }
        }
        catch (Exception)
        {
        }
    }
}
